### PR TITLE
fix: Event count fix for not allowed categories [DHIS2-7712]

### PR DESCRIPTION
**Pull request by maikelarabori**
Stop counting events related to the categories that users don't have access.
Porting it to 2.34. Was already ported to the other versions.

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/security/DefaultAnalyticsSecurityManager.java [hunk 1]
```diff
@@ -96,6 +99,28 @@ public DefaultAnalyticsSecurityManager( DataApprovalLevelService approvalLevelSe
     // AnalyticsSecurityManager implementation
     // -------------------------------------------------------------------------
 
+    /**
+     * Will remove/exclude, from DataQueryParams, any category option that the
+     * current user is not allowed to read.
+     *
+     * @param programCategories the categories related to this program.
+     */
+    void excludeNonAuthorizedCategoryOptions( final List<Category> programCategories )
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/security/DefaultAnalyticsSecurityManager.java]: Add modifier (private)

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/JdbcEventAnalyticsManager.java [hunk 2]
```diff
@@ -519,6 +532,56 @@ else if ( params.isOrganisationUnitMode( OrganisationUnitSelectionMode.CHILDREN
...
+     * authorized category options.
+     *
+     * @see org.hisp.dhis.analytics.AnalyticsSecurityManager#decideAccess(DataQueryParams)
+     *
+     * @param programCategories the list of program categories containing the list
+     *        of authorized category options for the current user.
+     * @return the sql statement in the format:
+     *          "and ax."mEXqxV2KIUl" in ('qNqYLugIySD') or ax."r7NDRdgj5zs" in ('qNqYLugIySD') "
+     */
+    String queryOnlyAllowedCategoryOptionEvents( final List<Category> programCategories )
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/JdbcEventAnalyticsManager.java]: Add modifier (private)

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/JdbcEventAnalyticsManager.java [hunk 3]
```diff
@@ -519,6 +532,56 @@ else if ( params.isOrganisationUnitMode( OrganisationUnitSelectionMode.CHILDREN
...
+    {
+        boolean andFlag = true;
+        final StringBuilder query = new StringBuilder();
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/JdbcEventAnalyticsManager.java]: We have a util class `SqlHelper`, you can add a "andOr" method there to encapsulate the and/or logic.

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/security/DefaultAnalyticsSecurityManager.java (hunk 1 above)
Comment (maikelarabori) [reply dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/security/DefaultAnalyticsSecurityManager.java]: We decided to leave as is to favour testability.

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/JdbcEventAnalyticsManager.java (hunk 2 above)
Comment (maikelarabori) [reply dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/JdbcEventAnalyticsManager.java]: We decided to leave as is to favour testability.

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/JdbcEventAnalyticsManager.java (hunk 3 above)
Comment (maikelarabori) [reply dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/JdbcEventAnalyticsManager.java]: Changes applied. Waiting Trevis to build.

**Comment by stian-sandvold** [comment]
Merging this since all of @larshelge comments have been resolved and Luciano has approved.

**Pull request merged**